File: apps/ios/Eno/Core/RecentStore.swift
import Foundation

// Device-local recency stores (guest-friendly, mirroring the web's localStorage
// recents): recently VIEWED listing ids feed the home rail via /api/listings?ids=,
// recent SEARCH terms feed the search screen's empty state. Auth sync can layer
// on later without changing callers.
enum RecentStore {
    private static let viewedKey = "eno-recently-viewed"
    private static let searchKey = "eno-recent-searches"

    static func viewedIds() -> [String] {
        UserDefaults.standard.stringArray(forKey: viewedKey) ?? []
    }

    static func recordViewed(_ id: String) {
        var ids = viewedIds().filter { $0 != id }
        ids.insert(id, at: 0)
        UserDefaults.standard.set(Array(ids.prefix(12)), forKey: viewedKey)
    }

    static func searches() -> [String] {
        UserDefaults.standard.stringArray(forKey: searchKey) ?? []
    }

    static func recordSearch(_ term: String) {
        let t = term.trimmingCharacters(in: .whitespaces)
        guard t.count >= 2 else { return }
        var all = searches().filter { $0.caseInsensitiveCompare(t) != .orderedSame }
        all.insert(t, at: 0)
        UserDefaults.standard.set(Array(all.prefix(8)), forKey: searchKey)
    }

    static func clearSearches() {
        UserDefaults.standard.removeObject(forKey: searchKey)
    }
}

File: apps/android/app/src/main/java/vn/eno/native_/core/RecentStore.kt
package vn.eno.native_.core

import android.content.Context
import org.json.JSONArray

// Device-local recency (mirror of apps/ios RecentStore): recently VIEWED
// listing ids feed the home rail via /api/listings?ids=; recent SEARCH terms
// feed the search screen's empty state.
object RecentStore {
    private const val PREFS = "eno"
    private const val VIEWED = "eno-recently-viewed"
    private const val SEARCHES = "eno-recent-searches"

    private fun prefs(ctx: Context) = ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    private fun readList(ctx: Context, key: String): List<String> =
        runCatching {
            val arr = JSONArray(prefs(ctx).getString(key, "[]") ?: "[]")
            (0 until arr.length()).map { arr.getString(it) }
        }.getOrDefault(emptyList())

    private fun writeList(ctx: Context, key: String, list: List<String>) {
        val arr = JSONArray()
        list.forEach { arr.put(it) }
        prefs(ctx).edit().putString(key, arr.toString()).apply()
    }

    fun viewedIds(ctx: Context): List<String> = readList(ctx, VIEWED)

    fun recordViewed(ctx: Context, id: String) {
        val list = (listOf(id) + readList(ctx, VIEWED).filter { it != id }).take(12)
        writeList(ctx, VIEWED, list)
    }

    fun searches(ctx: Context): List<String> = readList(ctx, SEARCHES)

    fun recordSearch(ctx: Context, term: String) {
        val t = term.trim()
        if (t.length < 2) return
        val list = (listOf(t) + readList(ctx, SEARCHES).filter { !it.equals(t, ignoreCase = true) }).take(8)
        writeList(ctx, SEARCHES, list)
    }

    fun clearSearches(ctx: Context) = writeList(ctx, SEARCHES, emptyList())
}
